AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+//
+//  ExampleAudioEngineDevice.m
+//  AudioDeviceExample
+//
+//  Copyright © 2018 Twilio, Inc. All rights reserved.
+//
+
+#import "ExampleAudioEngineDevice.h"
+
+// We want to get as close to 10 msec buffers as possible because this is what the media engine prefers.
+static double const kPreferredIOBufferDuration = 0.01;
+// We will use stereo playback where available. Some audio routes may be restricted to mono only.
+static size_t const kPreferredNumberOfChannels = 1;
+// An audio sample is a signed 16-bit integer.
+static size_t const kAudioSampleSize = 2;
+static uint32_t const kPreferredSampleRate = 48000;
+
+typedef struct RendererAudioContext {
+    TVIAudioDeviceContext deviceContext;
+    size_t maxFramesPerBuffer;
+    AudioBufferList *bufferList;
+    void *renderBlock; // AVAudioEngineManualRenderingBlock
+} RendererAudioContext;
+
+typedef struct CapturerAudioContext {
+    TVIAudioDeviceContext deviceContext;
+    AudioBufferList *bufferList;
+    AudioUnit audioUnit;
+} CapturerAudioContext;
+
+// The RemoteIO audio unit uses bus 0 for ouptut, and bus 1 for input.
+static int kOutputBus = 0;
+static int kInputBus = 1;
+
+// This is the maximum slice size for RemoteIO (as observed in the field). We will double check at initialization time.
+static size_t kMaximumFramesPerBuffer = 1156;
+
+@interface ExampleAudioEngineDevice()
+
+@property (nonatomic, assign, getter=isInterrupted) BOOL interrupted;
+@property (nonatomic, assign) AudioUnit audioUnit;
+@property (nonatomic, assign) AudioBufferList audioBufferList;
```

Comment:
Is this buffer list used for capturing? It would be nice to name it more descriptively.